### ISSUE TITLE
Panic when reading nested map-values with a schema

### ISSUE THREAD

**Issue by folbricht-stripe** [parquet-go/parquet-go#38]
In the example test below, I'm trying to write/read a nested map-value using an equivalent but different schema. The struct used for writing/reading has the same field as the one that is used to generate the schema, but it uses different types. This currently fails with:

```
panic: reflect.Value.Convert: value of type parquet_test.SchemaMapValue cannot be converted to type parquet_test.MapValue
```

It's interesting this only happens when using a nested map-value. It's just fine in a slice or regular structs. Using struct as map-value is fine too, it's just when nesting is involved.

While the example below seems contrived since the `SchemaMessage` technically isn't necessary, I have a use-case that involves generating a new type using reflection which is then used with `parquet.SchemaOf()` to generate a schema that runs into the same issue.

```go
func TestMapValue(t *testing.T) {
	// Main struct used to mashal/unmarshal data.
	type NestedStruct struct {
		Val string
	}
	type MapValue struct {
		Nested NestedStruct
	}
	type Message struct {
		TestMap map[string]MapValue
	}

	// Clone of the main struct used exclusively to generate the schema.
	type SchemaNestedStruct struct {
		Val string
	}
	type SchemaMapValue struct {
		Nested SchemaNestedStruct
	}
	type SchemaMessage struct {
		TestMap map[string]SchemaMapValue
	}

	testKey, testValue := "test-key", "test-value"
	in := Message{
		TestMap: map[string]MapValue{
			"test-key": {
				Nested: NestedStruct{
					Val: testValue,
				},
			},
		},
	}

	var f bytes.Buffer

	// Generate a schema for reading/writing using an exact clone of data.
	schema := parquet.SchemaOf(SchemaMessage{})

	pw := parquet.NewGenericWriter[Message](&f, schema)
	_, err := pw.Write([]Message{in})
	if err != nil {
		t.Fatal(err)
	}

	err = pw.Close()
	if err != nil {
		t.Fatal(err)
	}

	pr := parquet.NewGenericReader[*Message](bytes.NewReader(f.Bytes()), schema)

	out := make([]*Message, 1)
	_, err = pr.Read(out)
	if err != nil {
		t.Fatal(err)
	}
	pr.Close()
	if want, got := testValue, out[0].TestMap[testKey].Nested.Val; want != got {
		t.Error("failed to read map value")
	}
}
```

I haven't yet tried to debug this further myself yet. Any pointers or suggestions for things to try out are appreciated.

**Comment by gernest** [comment]
@folbricht-stripe  thanks for using this package.

I think you are giving the `GenericReader` the wrong schema to work with.  Here is a working code snippet of your test.

```go
func TestMapValue(t *testing.T) {
	// Main struct used to mashal/unmarshal data.
	type NestedStruct struct {
		Val string
	}
	type MapValue struct {
		Nested NestedStruct
	}
	type Message struct {
		TestMap map[string]MapValue
	}

	// Clone of the main struct used exclusively to generate the schema.
	type SchemaNestedStruct struct {
		Val string
	}
	type SchemaMapValue struct {
		Nested SchemaNestedStruct
	}
	type SchemaMessage struct {
		TestMap map[string]SchemaMapValue
	}

	testKey, testValue := "test-key", "test-value"
	in := Message{
		TestMap: map[string]MapValue{
			"test-key": {
				Nested: NestedStruct{
					Val: testValue,
				},
			},
		},
	}

	var f bytes.Buffer

	// Generate a schema for reading/writing using an exact clone of data.
	schema := parquet.SchemaOf(SchemaMessage{})

	pw := parquet.NewGenericWriter[Message](&f, schema)
	_, err := pw.Write([]Message{in})
	if err != nil {
		t.Fatal(err)
	}

	err = pw.Close()
	if err != nil {
		t.Fatal(err)
	}

	pr := parquet.NewGenericReader[*Message](bytes.NewReader(f.Bytes()), parquet.SchemaOf(Message{}))

	out := make([]*Message, 1)
	_, err = pr.Read(out)
	if err != nil {
		t.Fatal(err)
	}
	pr.Close()
	if want, got := testValue, out[0].TestMap[testKey].Nested.Val; want != got {
		t.Error("failed to read map value")
	}
}
```

Since from parquet perspective both are the same. We just need to give `GenericReader` correct hints on what to work with.

The change is essentially on this line 

**before**
```go
	pr := parquet.NewGenericReader[*Message](bytes.NewReader(f.Bytes()), schema)
```

**after**
```go
	pr := parquet.NewGenericReader[*Message](bytes.NewReader(f.Bytes()), parquet.SchemaOf(Message{}))
```

Does this resolve your issue ?

**Comment by folbricht-stripe** [comment]
Unfortunately that's not really an option in my case. I'm reading directly into a protobuf message (which doesn't have `parquet` tags, only `json` tags). And the schema is generated from a type that's a clone of the proto message with its json tags replaced. So in this line, the two types would have to be different:

```go
	pr := parquet.NewGenericReader[*ProtoMessage](bytes.NewReader(f.Bytes()), parquet.SchemaOf(CloneOfProtoMessage{}))
```

Thinking about it a bit more, the whole issue would be solved if parquet-go was able to use `json:` tags. Would you be opposed an implementation of `parquet.SchemaOf()` that can use json-tags for field names? I could try to PR something for that.

**Comment by fpetkovski** [comment]
I can think of two potential solutions that can get you unblocked in the short term, without json tags support:

* Copy the protobuf data into your custom type manually, just like you are copying the struct and schema.
* Use `parquet.NewSchema` to generate the schema manually. For the example above, you can do it as follows:

```
schema := parquet.NewSchema("Message", parquet.Group{
    "TestMap": parquet.Group{
        "key_value": parquet.Group{
            "key": parquet.String(),
            "value": parquet.Group{
                "Nested": parquet.Group{
                    "Val": parquet.String(),
                },
            },
        },
    },
})
```

Using `json` tags might create all kinds of edge cases, so I don't know if it's the best way to go about solving this. We might want a more ergonomic way to create custom schemas which is less verbose and harder to get wrong.

**Comment by folbricht-stripe** [comment]
I can definitely try using `parquet.NewSchema`, but haven't had any luck getting that to work either. 

```go
schema := parquet.NewSchema("Message", parquet.Group{
    "TestMap": parquet.Group{
        "key_value": parquet.Group{
            "key": parquet.String(),
            "value": parquet.Group{
                "Nested": parquet.Group{
                    "Val": parquet.String(),
                },
            },
        },
    },
})
```

Fails with `panic: reflect: call of reflect.Value.MapIndex on struct Value`. Maybe there's an issue in the schema but I can't see it. Looks very much like the one generated by `parquet.SchemaOf()`

**Comment by fpetkovski** [comment]
Interesting, the test case passes for me. Here is the entire snippet 

```go
func TestMapValue(t *testing.T) {
	// Main struct used to mashal/unmarshal data.
	type NestedStruct struct {
		Val string
	}
	type MapValue struct {
		Nested NestedStruct
	}
	type Message struct {
		TestMap map[string]MapValue
	}

	schema := parquet.NewSchema("Message", parquet.Group{
		"TestMap": parquet.Group{
			"key_value": parquet.Group{
				"key": parquet.String(),
				"value": parquet.Group{
					"Nested": parquet.Group{
						"Val": parquet.String(),
					},
				},
			},
		},
	})

	testKey, testValue := "test-key", "test-value"
	in := Message{
		TestMap: map[string]MapValue{
			"test-key": {
				Nested: NestedStruct{
					Val: testValue,
				},
			},
		},
	}

	var f bytes.Buffer

	pw := parquet.NewGenericWriter[Message](&f, schema)
	_, err := pw.Write([]Message{in})
	if err != nil {
		t.Fatal(err)
	}

	err = pw.Close()
	if err != nil {
		t.Fatal(err)
	}

	pr := parquet.NewGenericReader[*Message](bytes.NewReader(f.Bytes()), parquet.SchemaOf(Message{}))

	out := make([]*Message, 1)
	_, err = pr.Read(out)
	if err != nil {
		t.Fatal(err)
	}
	pr.Close()
	if want, got := testValue, out[0].TestMap[testKey].Nested.Val; want != got {
		t.Error("failed to read map value")
	}
	fmt.Println(out[0].TestMap[testKey].Nested.Val)
}
```

Does that work for you?

**Comment by folbricht-stripe** [comment]
The problem is the reader, you should see it when you change
```go
pr := parquet.NewGenericReader[*Message](bytes.NewReader(f.Bytes()), parquet.SchemaOf(Message{}))
```
to
```go
pr := parquet.NewGenericReader[*Message](bytes.NewReader(f.Bytes()), schema)
```

**Comment by fpetkovski** [comment]
I see, the problem is that `parquet.Group` is a `Map` type and it assumes that the go type will be a map. When creating a schema using `parquet.SchemaOf`, the underlying type will be a `parquet.structNode` whose `Value` implementation correctly assumes that you are reconstructing values into a struct. 

I don't know if there is a good way to solve this besides exporting `structNodeOf` so that it can be used in client libraries. Any thoughts @gernest?

**Comment by gernest** [comment]
I agree on @fpetkovski  regarding `json` tags. Lots of edge cases to deal with so I don't see we can come up with a solution for that.

Back to this issue.

@folbricht-stripe  I think you are confusing me. Can we focus on one issue at a time. As far as I understand, the test case you provided was addressed by both of us.

Correct me If I'm wrong.

- You had an issue, created a test case for it . The test case was resolved.
- You then used a different data from the one you provided on the test case (protobuf ,json stuff) It didn't work for you.

If I'm right, the second bullet is a different issue. I suggest 

- if my solution doesn't work with the data you provided then we can keep this issue open, you can give me more info to work with to resolve it.
- If you need `json`  or `protobuf`  support please open a separate ticket for we will see what we can do.

**Comment by folbricht-stripe** [comment]
Sorry, but the original issue wasn't really resolved. The problem is that it is not possible to use the reader with a schema not made from exactly the same type. The following always fails if the data contains a map with a nested value, regardless of if the schema was made from a type with `SchemaOf()`, or created with `NewSchema()`:

```go
parquet.NewGenericReader[A](r, parquet.SchemaOf(B))
```

```go
parquet.NewGenericReader[A](r, parquet.NewSchema(...))
```

Both of the above will cause a panic.

The solution suggested was to use the same type but that doesn't address the problem (and isn't an option in my use-case).

```go
parquet.NewGenericReader[A](r, parquet.SchemaOf(A))
```

**Comment by gernest** [comment]
Did I misread the title of the issue ?

> Panic when reading nested map-values with a schema #38

This is title

I am 100% certain the example you provided is falling because of the schema you passed to `GenericReader` was wrong.

> The problem is that it is not possible to use the reader with a schema not made from exactly the same type.

I can be wrong but this is not what is described in this issue.  Here is part of your description 

```
In the example test below, I'm trying to write/read a nested map-value using an equivalent but different schema. The struct used for writing/reading has the same field as the one that is used to generate the schema, but it uses different types. This currently fails with:
```

You can  close this and open a new issue with new details and we will look into it.

**Comment by folbricht-stripe** [comment]
I'll clarify in a new issue, going to use the same example, but creating the schema with `NewSchema()` instead of `SchemaOf()` which highlights the problem a little better.
